### PR TITLE
[Bugfix][Model] GR00T-N1.7: honor processor_config image pipeline (use_albumentations / crop_fraction / letter_box_transform) to match Isaac-GR00T

### LINKED ISSUE #7081: [Bug]: GR00T-N1.7 eval image preprocessing ignores processor_config (use_albumentations, crop_fraction, letter_box_transform) and diverges from Isaac-GR00T
## Your current environment

GB200 (aarch64), vLLM-Omni `10701e18` (main), vLLM 0.28.0, torch 2.13.0+cu130, transformers 5.14.1.
Reference: Isaac-GR00T `51d4c89` (torch 2.11.0+cu130, transformers 4.57.6, albumentations 1.4.18, opencv 5.0.0.93).

## Your code version

`10701e18`

## 🐛 Describe the bug

`Gr00tN1d7Processor` (`vllm_omni/diffusion/models/gr00t/modeling/processing_gr00t_n1d7.py`) rebuilds the GR00T-N1.7 eval image transform in torchvision and swallows the checkpoint's `use_albumentations` / `letter_box_transform` keys in `**kwargs`. Both public checkpoints (`nvidia/GR00T-N1.7-3B`, `nvidia/GR00T-N1.7-DROID`) ship

```json
"use_albumentations": true, "letter_box_transform": false, "crop_fraction": 0.95,
"shortest_image_edge": 256, "image_target_size": [256, 256], "image_crop_size": [230, 230]
```

so Isaac-GR00T (what the checkpoints were trained with) and vLLM-Omni preprocess differently:

| step | Isaac-GR00T eval (`use_albumentations=True`) | vLLM-Omni main |
|---|---|---|
| letterbox | only if `letter_box_transform` (False here) | unconditional `LetterBoxTransform()` |
| resize | `SmallestMaxSize(256, INTER_AREA)` | `Resize([256, 256])`, bilinear |
| crop | `FractionalCenterCrop(0.95)`: 256 -> **243** | `CenterCrop([230, 230])`: fixed **230** |

230/256 = 0.898 vs 0.95: the model sees a narrower field of view than it was trained on, resampled with a different filter.

### Reproduction (no Isaac-GR00T needed; `cv2` ships with vLLM)

```python
import cv2, numpy as np, torch
from vllm_omni.diffusion.models.gr00t.modeling.processing_gr00t_n1d7 import _build_eval_image_transform

frame = (np.indices((256, 256)).sum(0) % 256).astype(np.uint8)[..., None].repeat(3, -1)
ours = _build_eval_image_transform([256, 256], [230, 230])(frame)   # main: 230 crop, bilinear
crop = frame[6:249, 6:249]                                          # Isaac: int(256 * 0.95) = 243, centred
isaac = torch.from_numpy(cv2.resize(crop, (256, 256), interpolation=cv2.INTER_AREA)).permute(2, 0, 1)
print((ours.int() - isaac.int()).abs().max())   # main: tensor(230) -- out of 255
```

### Measured impact

Isaac-GR00T pins `transformers==4.57.3` / `torch==2.9.0` while vLLM-Omni main needs transformers 5.x, so the two cannot share an interpreter. They were run as two processes on one GB200 node against `nvidia/GR00T-N1.7-DROID`, exchanging tensors through disk: the same observation file (one non-zero 256x256 frame per view), and the reference's own initial denoising noise captured and injected into the vLLM-Omni run so the ODE starts from the same point.

| tensor | main vs Isaac-GR00T | bitwise equal |
|---|---|---|
| `pixel_values` | **1.945** | no |
| `backbone_features` | 16.78 | no |
| `action_pred` (normalized) | 0.979 | no |
| decoded `eef_9d` / `gripper_position` / `joint_position` | 0.148 / 0.489 / 0.107 | no |
| `input_ids`, `attention_mask`, `image_grid_thw`, `state`, `embodiment_id`, `state_features` | 0 | yes |

Every model input except `pixel_values` is bitwise equal, which places the divergence in the image pipeline.

The downstream rows should be read with one caveat: because the two sides must run different torch/transformers builds, they carry a bf16 floor unrelated to preprocessing. Re-running the same comparison with a candidate fix that makes `pixel_values` bitwise equal leaves `backbone_features` at 1.375 and the decoded actions at 0.004-0.009 — so that floor is roughly an order of magnitude below the numbers above, and the `pixel_values` row itself is unaffected by it either way.

### Why CI does not catch this

`tests/e2e/online_serving/test_gr00t_openpi_expansion.py` compares against Isaac-derived golden actions, but its observation is all-zero images (`build_openpi_droid_observation`). An all-zero frame is a fixed point of every resize/crop recipe, so both pipelines produce identical pixels and the test passes either way.

### Relation to the rollout discussion on #3798

During review of #3798 a MolmoSpaces rollout returned [`success=0/10` on FrankaPickDroidMiniBench](https://github.com/vllm-project/vllm-omni/pull/3798#issuecomment-4535330850), and [the same failure reproduced on the official Isaac-GR00T codebase](https://github.com/vllm-project/vllm-omni/pull/3798#issuecomment-4543771096), which reasonably [led to attributing it to the checkpoint not being post-trained for that benchmark](https://github.com/vllm-project/vllm-omni/pull/3798#issuecomment-4550826983) rather than to the integration. That inference is sound for what it was testing — if the reference implementation fails too, the port is exonerated *for that experiment*.

The gap it leaves is that a rollout is coarse: it mixes model capability, benchmark fit and sampling randomness into one pass/fail, and it cannot separate a preprocessing difference from any of them. A layer-wise comparison on identical non-zero inputs — which is what this issue reports — was never run, and it does isolate the preprocessing.

One detail from that thread may be worth revisiting: the reported symptom was that [the gripper never closes, with `gripper_position` staying near zero](https://github.com/vllm-project/vllm-omni/pull/3798#issuecomment-4543771096). In the measurements above `gripper_position` is the decoded channel with the largest deviation (0.489 absolute, 0.979 relative). **I have not shown these are connected** — I have not re-run that benchmark with the preprocessing fixed, and since Isaac-GR00T failed it too, the post-training explanation may well be the whole story. It seems worth checking rather than assuming.

Related: #3798 (the port). Isaac-GR00T's `image_augmentations.py` has carried the albumentations recipe since the N1.7 release (23ace64).

### PR DESCRIPTION
**Pull request by liangmenghuang**
## Purpose

Fixes #7081.

### What broke
`Gr00tN1d7Processor` ignores the image-pipeline keys of the checkpoint's `processor_config.json`: `use_albumentations` and `letter_box_transform` are swallowed by `**kwargs`, and the eval transform is always `LetterBox -> Resize(256, bilinear) -> CenterCrop(230) -> Resize(256)`. Both public N1.7 checkpoints (`nvidia/GR00T-N1.7-3B`, `nvidia/GR00T-N1.7-DROID`) ship `use_albumentations: true, crop_fraction: 0.95, shortest_image_edge: 256, image_crop_size: [230, 230], letter_box_transform: false`, which in Isaac-GR00T means `SmallestMaxSize(256, INTER_AREA) -> FractionalCenterCrop(0.95) -> SmallestMaxSize(256, INTER_AREA)`: a 243 px crop (not 230; 230/256 = 0.898 vs 0.95), area resampling (not bilinear), no letterbox. The model gets out-of-distribution pixels.

Measured on GB200 with the DROID checkpoint against Isaac-GR00T 51d4c89, on one non-zero frame per view and with the reference's own initial denoising noise injected so the ODE starts from the same point: `pixel_values` max abs diff 1.945, `backbone_features` 16.78, decoded actions 0.107-0.489, while every other model input (`input_ids`, `attention_mask`, `image_grid_thw`, `state`, `embodiment_id`) is already bitwise equal — which places the divergence in the image pipeline. Full numbers, including the arm of the same experiment run on this branch, are under Test Result.

### Repro (main, no Isaac-GR00T needed; cv2 ships with vLLM)
```python
import cv2, numpy as np, torch
from vllm_omni.diffusion.models.gr00t.modeling.processing_gr00t_n1d7 import _build_eval_image_transform

frame = (np.indices((256, 256)).sum(0) % 256).astype(np.uint8)[..., None].repeat(3, -1)  # any HWC uint8 frame
crop = frame[6:249, 6:249]                                     # Isaac: int(256 * 0.95) = 243, centred
isaac = torch.from_numpy(cv2.resize(crop, (256, 256), interpolation=cv2.INTER_AREA)).permute(2, 0, 1)

# What the checkpoint's processor_config actually asks for -- only reachable on this branch:
fixed = _build_eval_image_transform(
    [256, 256], [230, 230], use_albumentations=True, shortest_image_edge=256, crop_fraction=0.95
)(frame)
print((fixed.int() - isaac.int()).abs().max())   # this PR: tensor(0)

# The recipe main always runs (and this branch still runs when the flag is off):
ours = _build_eval_image_transform([256, 256], [230, 230])(frame)
print((ours.int() - isaac.int()).abs().max())    # tensor(230) -- out of 255
```
`_build_eval_image_transform` keeps its positional signature, so the second call behaves the same on main and on this branch; the flags are keyword-only and come from `processor_config.json`.

### Root cause
The port re-implemented Isaac-GR00T's `build_image_transformations` (torchvision recipe) with the letterbox hard-wired on, and never ported `build_image_transformations_albumentations`, the recipe selected by `use_albumentations=True`. The processor `__init__` dropped both selector keys as "deprecated training-only keys".

### Fix
- `Gr00tN1d7Processor.__init__` takes `use_albumentations` and `letter_box_transform` explicitly (defaults `False`, same as Isaac-GR00T) instead of dropping them; `save_pretrained` persists them so save/load round-trips.
- `_build_eval_image_transform` selects the recipe like Isaac's processor does: `use_albumentations=True` -> new `AlbumentationsEvalImageTransform`, a cv2/numpy transcription of Isaac-GR00T's eval branch and albumentations 1.4.18 (`SmallestMaxSize` rounding, `FractionalCenterCrop` truncation, `LetterBoxPad`), including the `crop_fraction`/`shortest_image_edge` `None` fallbacks. No albumentations dependency; `cv2` is already there via vLLM's `opencv-python-headless`, imported lazily inside the helpers like `cosmos3/transfer.py`.
- The torchvision recipe now gates `LetterBoxTransform` on `letter_box_transform`, as Isaac-GR00T does (no-op for square frames).
- The per-frame contract at `_get_vlm_inputs` (HWC uint8 in, `(C, H, W)` uint8 tensor out) is unchanged, so the call site is untouched.

### Behavior change
- Checkpoints with `use_albumentations: true` (both public N1.7 checkpoints) now get the pixels they were trained on; outputs change (correctness fix).
- Checkpoints with `use_albumentations: false` or without the key keep the torchvision recipe; the only difference is that letterboxing now follows `letter_box_transform` (absent -> `False`, Isaac's default), which is a no-op for square frames.
- `state_dropout_prob` / `extra_augmentation_config` remain training-only and are still ignored.

Not touched: #6568 / #6450 (pipeline_gr00t.py). #5028 changes the same file, but its `AutoProcessor.register(Gr00tN1d7Config, ...)` fix is already on main via #5042.

## Test Plan

New CPU test `tests/diffusion/models/gr00t/test_processing_eval_image_transform.py` (`core_model` + `cpu`, runs in "Simple · Diffusion Test"):
- recipe selection from the flags, `None` fallbacks, geometry/rounding rules, contract errors, letterbox gating on both recipes, `_get_vlm_inputs` end-to-end with the new transform, `save_pretrained` -> `from_pretrained` round-trip, legacy configs without the flags;
- bit-exact goldens: sha256 of the `(C, H, W)` uint8 output of **Isaac-GR00T's own** eval transform (`build_image_transformations_albumentations(...)[1]` through `apply_with_replay`, Isaac-GR00T 51d4c89 / albumentations 1.4.18 / opencv 5.0.0.93) on 4 deterministic integer-generated frames x letterbox on/off. The non-square frame is 512x768 so the INTER_AREA downscales are integer factors (exact integer sums in OpenCV): the goldens do not depend on the CPU architecture. Generating them also cross-checked Isaac-GR00T, this branch's code and an independent transcription for bitwise agreement on every case, and the torchvision branch against Isaac's torchvision eval transform. Regenerating a digest needs Isaac-GR00T installed, whose transformers pin conflicts with this repo's, so it cannot run in CI; the test module's docstring states the procedure.

```bash
pytest -sv tests/diffusion/models/gr00t -m 'core_model and cpu'
pre-commit run --all-files
```

Processor-level parity (Isaac-GR00T processor vs patched processor, transformers 4.57.6 environment, DROID checkpoint, 2 seeds x batch 1/2): `pixel_values`, `input_ids`, `image_grid_thw`, `attention_mask`, `state`, `backbone_features`, `absolute_action` all bitwise equal, also through the real serving path.

**vLLM Version:** 0.28.0

**vLLM-Omni Commit:** 10701e18 (upstream/main, branched here)

## Test Result

```
$ pytest -sv tests/diffusion/models/gr00t -m 'core_model and cpu'
21 passed, 22 warnings in 3.68s
```
(17 new cases plus the 4 pre-existing pipeline tests. On main the new module fails to import
`AlbumentationsEvalImageTransform`; against main's recipe the pinned Isaac-GR00T goldens differ by
up to 255 gray levels on the checkerboard frame, 247 on the noise frame, and the non-square frame
even differs in shape, 256x256 instead of 256x383.)

### End to end on GB200

Isaac-GR00T pins `transformers==4.57.3` / `torch==2.9.0` while vLLM-Omni main needs transformers 5.x, so the two cannot share an interpreter. They were run as two processes on one GB200 node against `nvidia/GR00T-N1.7-DROID`, exchanging tensors through disk: the same observation file (one non-zero 256x256 frame per view — a zero frame is a fixed point of every resize/crop recipe and would show nothing), and the reference's own initial denoising noise captured and injected into the vLLM-Omni run. The serving side drives `Gr00tPolicy.get_action` and imports nothing but `vllm_omni`. Both arms use the same reference dump and differ only by `git checkout`:

| tensor | main | this branch |
|---|---|---|
| `pixel_values` | 1.945 | **0 (bitwise)** |
| `input_ids`, `attention_mask`, `image_grid_thw`, `state`, `embodiment_id`, `state_features` | 0 | 0 |
| `backbone_features` | 16.78 | 1.375 |
| `action_pred` (normalized) | 0.979 | 0.031 |
| decoded `eef_9d` / `gripper_position` / `joint_position` | 0.148 / 0.489 / 0.107 | 0.004 / 0.006 / 0.009 |

The processor reports its recipe as `Compose` on main and `AlbumentationsEvalImageTransform` on this branch, so the switch is observable rather than inferred.

`pixel_values` — the tensor this PR is about — goes to bitwise equality. The downstream rows do not, and should not be read as if they could: the two sides necessarily run different torch/transformers builds, and that difference alone is worth roughly the residual in the right-hand column. Both arms carry that same floor, so the honest claim is the contrast between the columns, not the absolute value of either. A same-process end-to-end bitwise check is not reachable while the dependency pins conflict.

### CODE REVIEW

**Comment by liangmenghuang** [comment]
Self-review.

**What changed**: `Gr00tN1d7Processor` now reads `use_albumentations` / `letter_box_transform` from `processor_config.json` instead of dropping them into `**kwargs`, and `_build_eval_image_transform` selects the recipe the way Isaac-GR00T's processor does. Two files: the processor and a new CPU test module.

**What I checked**:

- **Behavior gating** — checkpoints without the keys keep the previous torchvision recipe (defaults `False`, matching Isaac-GR00T); only the letterbox now follows `letter_box_transform`, a no-op for square frames. Both public N1.7 checkpoints ship `use_albumentations: true`, so they change behavior — that is the correctness fix.
- **Call-site contract** — `_get_vlm_inputs` still gets HWC uint8 in / `(C, H, W)` uint8 tensor out, so the single call site is untouched.
- **Round-trip** — both flags are persisted by `save_pretrained`; there is a `save_pretrained` -> `from_pretrained` test.
- **No new dependency** — `cv2` comes from vLLM's `opencv-python-headless` and is imported lazily inside the helpers.
- **Reachability of the `None` fallbacks** — `image_target_size` / `image_crop_size` default to `None` in the processor `__init__`, so the guards in `_build_eval_image_transform` are live, not defensive dead code.
- **Tests** — `pytest tests/diffusion/models/gr00t -m 'core_model and cpu'` -> 21 passed (17 new + 4 pre-existing). Goldens are sha256 of Isaac-GR00T's own eval transform output; the non-square frame is 512x768 so every INTER_AREA downscale is an integer factor and the goldens are architecture-independent.
- **Pre-commit** — full hook set passes locally (SPDX, forbidden imports, no new `torch.cuda`, mypy-3.10, test CI marks, ruff, typos).
- **End to end** — two-process comparison against Isaac-GR00T on GB200, numbers in the description. `pixel_values` reaches bitwise equality; the downstream residual is the different torch/transformers builds the two sides must run, and it is present in both arms.

**What I did not check**: whether this changes any rollout benchmark outcome. The `gripper_position` observation in #7081 is flagged there as unverified.

**Comment by princepride** [comment]
@tzhouam PTAL

File: tests/diffusion/models/gr00t/test_processing_eval_image_transform.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM-Omni project
+"""GR00T-N1.7 eval image transform: chosen from ``processor_config.json`` and bit-exact with Isaac-GR00T.
+
+``GOLDEN`` holds sha256 digests of the ``(C, H, W)`` uint8 output of Isaac-GR00T's own eval transform
+(``build_image_transformations_albumentations(...)[1]`` applied through ``apply_with_replay``) for the
+synthetic frames built below, at Isaac-GR00T 51d4c89 / albumentations 1.4.18 / opencv 5.0.0.93. The frame
+generators are integer-only so they are identical on every platform; keep them byte-identical to the
+regeneration script attached to the PR, which also cross-checks this module against Isaac-GR00T.
```
Comment (timzsu) [inline tests/diffusion/models/gr00t/test_processing_eval_image_transform.py]: Where is the "regeneration script" mentioned in the docstring? If it is removed from the PR, I suggest cleaning up the docstring as well.

**Review by timzsu — APPROVED** [review]
This PR adds support for two transform options that my initial porting missed. The change is easy to read through and reproduction plus end-to-end precision numbers are also provided. LGTM overall. 

I have left a minor comment on the docstring which seems to reference a non-existent file. PTAL when you have time.

File: tests/diffusion/models/gr00t/test_processing_eval_image_transform.py (hunk 1 above)
Comment (liangmenghuang) [reply tests/diffusion/models/gr00t/test_processing_eval_image_transform.py]: Sorry, misleading wording — removed the "Golden regeneration script" block from the PR description. `eaeaae80` cleans up the docstring so it states the regeneration procedure inline; `GOLDEN` and the tests are unchanged.

**Review by princepride — APPROVED** [review]
Lgtm
